src/cli/require-templates.js

```diff
@@ -8,99 +8,52 @@ export default function requireTemplates(huron, templates, sections) {
-      // (i.e. the hbs template or the JSON object) and not the the object pair.
-      // The following callback was a suggestion from Owen:
-      //
-      // `renderCallback = ${huron.templates.callback.stringify()}`,
-      `templateReplaceCallback = cb;`,
-    `}`,
-    `let templateReplaceCallback = null`,
-    `let renderCallback = null`,
-    `if (module.hot) {`,
-      `module.hot.accept(`,
-        `[${templatePathArray}],`,
-        `(update) => {`,
-          `let updatedModules = Object.keys(update);`,
-          `if (updatedModules.length) {`,
-            `let templateModules = update[updatedModules[0]];`,
-            `if (templateModules.length) {`,
-              `templateModules.forEach( (templateKey) => {`,
-                `let template = __webpack_require__(templateKey);`,
-                `templateReplaceCallback(template, templates);`,
-              `});`,
-            `}`,
-          `}`,
-        `}`,
-      `);`,
-    `}`
-  ];
-
-  // Add the function to render the templates
-  prependScript.push(
-    `export const templateCallback = ` + huron.templates.callback.toString() + `;`
-  );
+  let prependScript = `export const modules = {};
+export const templates = ${JSON.stringify(templateObj)};
+export function addCallback(cb) {
+  templateReplaceCallback = cb;
+}
+let templateReplaceCallback = null;
+let assets = require.context(
+  '${path.join(cwd, huron.root, huron.output)}',
+  true,
+  /\.(html|json|${huron.templates.extension.replace('.', '')})/
+);
 
-  // Generate templates object using template IDs as keys.
-  // This will require the files, or create an object with keys and
-  // the required files.
-  for (let template in templateObj) {
-    prependScript.push(
-      `templates['${template}'] =`
-    );
+assets.keys().forEach(function(key) {
+  modules[key] = assets(key);
+});
 
-    // @todo this seems pretty duplicative of the code that adds
-    // the stuff for hot module reloading above and should be refactored.
-    if ('object' === typeof templateObj[template]) {
-      prependScript.push(
-      `{`
-      );
-      for (let type in templateObj[template]) {
-         prependScript.push(`${type}: require('${templateObj[template][type]}'),`);
-      }
-      prependScript.push(
-      `};`
-      );
-    } else {
-      prependScript.push(
-        `require('${templateObj[template]}');`
+if (module.hot) {
+  module.hot.accept(
+    assets.id,
+    () => {
+      const newAssets = require.context(
+        '${path.join(cwd, huron.root, huron.output)}',
+        true,
+        /\.(html|json|${huron.templates.extension.replace('.', '')})/
       );
+      const newModules = newAssets.keys()
+        .map((key) => {
+          return [key, newAssets(key)];
+        })
+        .filter((newModule) => {
+          return modules[newModule[0]] !== newModule[1];
+        });
+
+      newModules.forEach((module) => {
+        modules[module[0]] = module[1];
```

Comment:
`module[0]` = Relative require path (so something like `./partials/my-template.hbs`). `modules` is an object in which the keys are require paths like that, and the values are the modules themselves.
`module[1]` = the contents of the module itself.

So this line here will update both the module path (or filename, or whatever) and the module contents to make sure the whole list of modules is up-to-date.